gcloud/monitoring/timeseries.py

```diff
@@ -0,0 +1,706 @@
+            This can be the name of a resource label or metric label
+            (e.g., ``"instance_name"``), or the string ``"resource_type"``.
+
+        :type labels: list of strings, or None
+        :param labels: A list or tuple of label names to use for the dataframe
+            header. If more than one label name is provided, the resulting
+            dataframe will have a multi-level column header. Providing values
+            for both ``label`` and ``labels`` is an error.
+
+        :rtype: :class:`pandas.DataFrame`
+        :returns: A dataframe where each column represents one time series.
+        """
+        import pandas   # pylint: disable=import-error
+
+        if label is not None and labels is not None:
+            raise ValueError('Cannot specify both "label" and "labels".')
+        elif not (labels or labels is None):
+            raise ValueError('"labels" must be non-empty or None.')
+
+        columns = []
+        headers = []
+        for time_series in self:
+            pandas_series = pandas.Series(
+                data=[p.value for p in time_series.points],
+                index=[p.end_time for p in time_series.points],
+            )
+            columns.append(pandas_series)
+            headers.append(time_series._replace(points=None))
+
+        # Implement a smart default of using all available labels.
+        if label is None and labels is None:
+            resource_labels = set(itertools.chain.from_iterable(
+                header.resource.labels for header in headers))
+            metric_labels = set(itertools.chain.from_iterable(
+                header.metric.labels for header in headers))
+            labels = (['resource_type'] +
+                      _sorted_resource_labels(resource_labels) +
+                      sorted(metric_labels))
+
+        # Assemble the columns into a DataFrame.
+        dataframe = pandas.DataFrame(columns).T
+
+        # Convert the timestamp strings into a DatetimeIndex.
+        dataframe.index = pandas.to_datetime(dataframe.index)
+
+        # Build a column Index or MultiIndex from the label values. Do not
+        # include level names in the column header if the user requested a
+        # single-level header by specifying "label".
+        level_names = labels or None
+        label_keys = labels or [label]
+        dataframe.columns = pandas.MultiIndex.from_arrays(
+            [[header.labels.get(key, '') for header in headers]
+             for key in label_keys],
+            names=level_names)
+
+        # Sort the rows just in case (since the API doesn't guarantee the
+        # ordering), and sort the columns lexicographically.
+        return dataframe.sort_index(axis=0).sort_index(axis=1)
+
+    def copy(self):
+        """Copy the query object.
+
+        :rtype: :class:`Query`
+        :returns: The new query object.
+        """
+        new_query = copy.copy(self)
+        new_query._filter = copy.copy(self._filter)
```

Comment:
I'm not sure if you mean 1) instead of the pair of calls to `copy.copy()` or 2) instead of the second call.

1. I'll add a comment explaining that this code specifically does _not_ copy `_client`.
2. Doing a deep copy of `_filter` is unnecessary because it never mutates its attributes.